// @jupyterlab/application/lib/lab.js
"use strict";
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
Object.defineProperty(exports, "__esModule", { value: true });
const coreutils_1 = require("@jupyterlab/coreutils");
const docregistry_1 = require("@jupyterlab/docregistry");
const coreutils_2 = require("@phosphor/coreutils");
const frontend_1 = require("./frontend");
const mimerenderers_1 = require("./mimerenderers");
const shell_1 = require("./shell");
const status_1 = require("./status");
/**
 * JupyterLab is the main application class. It is instantiated once and shared.
 */
class JupyterLab extends frontend_1.JupyterFrontEnd {
    /**
     * Construct a new JupyterLab object.
     */
    constructor(options = { shell: new shell_1.LabShell() }) {
        super({ shell: options.shell || new shell_1.LabShell() });
        /**
         * The name of the JupyterLab application.
         */
        this.name = coreutils_1.PageConfig.getOption('appName') || 'JupyterLab';
        /**
         * A namespace/prefix plugins may use to denote their provenance.
         */
        this.namespace = coreutils_1.PageConfig.getOption('appNamespace') || this.name;
        /**
         * A list of all errors encountered when registering plugins.
         */
        this.registerPluginErrors = [];
        /**
         * The application busy and dirty status signals and flags.
         */
        this.status = new status_1.LabStatus(this);
        /**
         * The version of the JupyterLab application.
         */
        this.version = coreutils_1.PageConfig.getOption('appVersion') || 'unknown';
        this.restored = this.shell.restored
            .then(() => undefined)
            .catch(() => undefined);
        // Create an IInfo dictionary from the options to override the defaults.
        const info = Object.keys(JupyterLab.defaultInfo).reduce((acc, val) => {
            if (val in options) {
                acc[val] = JSON.parse(JSON.stringify(options[val]));
            }
            return acc;
        }, {});
        // Populate application info.
        this._info = Object.assign({}, JupyterLab.defaultInfo, info);
        // Populate application paths override the defaults if necessary.
        const defaultURLs = JupyterLab.defaultPaths.urls;
        const defaultDirs = JupyterLab.defaultPaths.directories;
        const optionURLs = (options.paths && options.paths.urls) || {};
        const optionDirs = (options.paths && options.paths.directories) || {};
        this._paths = {
            urls: Object.keys(defaultURLs).reduce((acc, key) => {
                if (key in optionURLs) {
                    const value = optionURLs[key];
                    acc[key] = value;
                }
                else {
                    acc[key] = defaultURLs[key];
                }
                return acc;
            }, {}),
            directories: Object.keys(JupyterLab.defaultPaths.directories).reduce((acc, key) => {
                if (key in optionDirs) {
                    const value = optionDirs[key];
                    acc[key] = value;
                }
                else {
                    acc[key] = defaultDirs[key];
                }
                return acc;
            }, {})
        };
        if (this._info.devMode) {
            this.shell.addClass('jp-mod-devMode');
        }
        // Add initial model factory.
        this.docRegistry.addModelFactory(new docregistry_1.Base64ModelFactory());
        if (options.mimeExtensions) {
            for (let plugin of mimerenderers_1.createRendermimePlugins(options.mimeExtensions)) {
                this.registerPlugin(plugin);
            }
        }
    }
    /**
     * The JupyterLab application information dictionary.
     */
    get info() {
        return this._info;
    }
    /**
     * The JupyterLab application paths dictionary.
     */
    get paths() {
        return this._paths;
    }
    /**
     * Register plugins from a plugin module.
     *
     * @param mod - The plugin module to register.
     */
    registerPluginModule(mod) {
        let data = mod.default;
        // Handle commonjs exports.
        if (!mod.hasOwnProperty('__esModule')) {
            data = mod;
        }
        if (!Array.isArray(data)) {
            data = [data];
        }
        data.forEach(item => {
            try {
                this.registerPlugin(item);
            }
            catch (error) {
                this.registerPluginErrors.push(error);
            }
        });
    }
    /**
     * Register the plugins from multiple plugin modules.
     *
     * @param mods - The plugin modules to register.
     */
    registerPluginModules(mods) {
        mods.forEach(mod => {
            this.registerPluginModule(mod);
        });
    }
}
exports.JupyterLab = JupyterLab;
/**
 * The namespace for `JupyterLab` class statics.
 */
(function (JupyterLab) {
    /* tslint:disable */
    /**
     * The layout restorer token.
     */
    JupyterLab.IInfo = new coreutils_2.Token('@jupyterlab/application:IInfo');
    /**
     * The default JupyterLab application info.
     */
    JupyterLab.defaultInfo = {
        devMode: coreutils_1.PageConfig.getOption('devMode').toLowerCase() === 'true',
        deferred: { patterns: [], matches: [] },
        disabled: { patterns: [], matches: [] },
        mimeExtensions: [],
        filesCached: coreutils_1.PageConfig.getOption('cacheFiles').toLowerCase() === 'true'
    };
    /**
     * The default JupyterLab application paths.
     */
    JupyterLab.defaultPaths = {
        urls: {
            base: coreutils_1.PageConfig.getOption('baseUrl'),
            defaultWorkspace: coreutils_1.PageConfig.getOption('defaultWorkspace'),
            notFound: coreutils_1.PageConfig.getOption('notFoundUrl'),
            page: coreutils_1.PageConfig.getOption('pageUrl'),
            public: coreutils_1.PageConfig.getOption('publicUrl'),
            settings: coreutils_1.PageConfig.getOption('settingsUrl'),
            themes: coreutils_1.PageConfig.getOption('themesUrl'),
            tree: coreutils_1.PageConfig.getOption('treeUrl'),
            workspaces: coreutils_1.PageConfig.getOption('workspacesUrl')
        },
        directories: {
            appSettings: coreutils_1.PageConfig.getOption('appSettingsDir'),
            schemas: coreutils_1.PageConfig.getOption('schemasDir'),
            static: coreutils_1.PageConfig.getOption('staticDir'),
            templates: coreutils_1.PageConfig.getOption('templatesDir'),
            themes: coreutils_1.PageConfig.getOption('themesDir'),
            userSettings: coreutils_1.PageConfig.getOption('userSettingsDir'),
            serverRoot: coreutils_1.PageConfig.getOption('serverRoot'),
            workspaces: coreutils_1.PageConfig.getOption('workspacesDir')
        }
    };
})(JupyterLab = exports.JupyterLab || (exports.JupyterLab = {}));
