// docz-theme-umi/es/components/ui/Playground/ActionsBar.js
import * as React from 'react';
import { useConfig } from 'docz';
import lighten from 'polished/lib/color/lighten';
import darken from 'polished/lib/color/darken';
import Maximize from 'react-feather/dist/icons/maximize';
import Minimize from 'react-feather/dist/icons/minimize';
import Refresh from 'react-feather/dist/icons/refresh-cw';
import Code from 'react-feather/dist/icons/code';
import styled, { css } from 'styled-components';
import { get as themeGet } from "../../../utils/theme";
import { CodeSandboxLogo } from "./CodeSandboxLogo";
import { ActionButton, ClipboardAction } from "../Editor/elements";
var borderColor = themeGet('colors.border');

var getActionsBg = function getActionsBg(p) {
  return p.theme.docz.mode === 'light' ? lighten(0.13, borderColor(p)) : darken(0.04, borderColor(p));
};

var Actions = styled.div.withConfig({
  displayName: "ActionsBar__Actions"
})(["display:flex;justify-content:flex-end;padding:0 5px;background:", ";"], getActionsBg);
var actionStyle = css(["padding:3px 10px;border-left:1px solid ", ";"], borderColor);
var Action = styled(ActionButton).withConfig({
  displayName: "ActionsBar__Action"
})(["", ";"], actionStyle);
var Clipboard = styled(ClipboardAction).withConfig({
  displayName: "ActionsBar__Clipboard"
})(["", ";"], actionStyle);
export var ActionsBar = function ActionsBar(_ref) {
  var showEditor = _ref.showEditor,
      code = _ref.code,
      fullscreen = _ref.fullscreen,
      codesandboxUrl = _ref.codesandboxUrl,
      onClickRefresh = _ref.onClickRefresh,
      onClickFullscreen = _ref.onClickFullscreen,
      onClickEditor = _ref.onClickEditor;
  var config = useConfig();
  var hasSandbox = Boolean(config.codeSandbox);
  return React.createElement(Actions, {
    withRadius: showEditor
  }, React.createElement(Action, {
    onClick: onClickRefresh,
    title: "Refresh playground"
  }, React.createElement(Refresh, {
    width: 15
  })), hasSandbox && React.createElement(Action, {
    as: "a",
    href: codesandboxUrl,
    target: "_blank",
    title: "Open in CodeSandbox"
  }, React.createElement(CodeSandboxLogo, {
    style: {
      height: '100%'
    },
    width: 15
  })), React.createElement(Clipboard, {
    content: code
  }), React.createElement(Action, {
    onClick: onClickFullscreen,
    title: fullscreen ? 'Minimize' : 'Maximize'
  }, fullscreen ? React.createElement(Minimize, {
    width: 15
  }) : React.createElement(Maximize, {
    width: 15
  })), React.createElement(Action, {
    onClick: onClickEditor,
    title: showEditor ? 'Close editor' : 'Show editor'
  }, React.createElement(Code, {
    width: 15
  })));
};
