// docz-theme-umi/es/components/ui/Table.js
import * as React from 'react';
import styled from 'styled-components';
import { get } from "../../utils/theme";
import { mq } from "../../styles/responsive";
var Wrapper = styled.div.withConfig({
  displayName: "Table__Wrapper"
})(["overflow-x:auto;padding:2px;margin-bottom:30px;", ";"], mq({
  marginBottom: [20, 40],
  maxWidth: ['calc(100vw - 40px)', 'calc(100vw - 80px)', '100%']
}));
var TableStyled = styled.table.withConfig({
  displayName: "Table__TableStyled"
})(["padding:0;table-layout:auto;box-shadow:0 0 0 1px ", ";background-color:transparent;border-spacing:0;border-collapse:collapse;border-style:hidden;border-radius:", ";font-size:14px;color:", ";", " & thead{color:", ";background:", ";}& thead th{font-weight:400;padding:20px 20px;&:nth-of-type(1){", ";}&:nth-of-type(2){", ";}&:nth-of-type(3){", ";}&:nth-of-type(4){", ";}&:nth-of-type(5){", ";}}& tbody td{padding:12px 20px;line-height:2;font-weight:200;}& tbody > tr{display:table-row;border-top:1px solid ", ";}", ";"], get('colors.border'), get('radii'), get('colors.tableColor'), mq({
  overflowX: ['initial', 'initial', 'initial', 'hidden'],
  display: ['block', 'block', 'block', 'table']
}), get('colors.theadColor'), get('colors.theadBg'), mq({
  width: ['20%', '20%', '20%', 'auto']
}), mq({
  width: ['10%', '10%', '10%', 'auto']
}), mq({
  width: ['10%', '10%', '10%', 'auto']
}), mq({
  width: ['10%', '10%', '10%', 'auto']
}), mq({
  width: ['20%', '20%', '20%', 'auto']
}), get('colors.border'), get('styles.table'));
export var Table = function Table(props) {
  return React.createElement(Wrapper, null, React.createElement(TableStyled, props));
};
