// docz-theme-umi/es/components/shared/Sidebar/Hamburger.js
import * as React from 'react';
import styled from 'styled-components';
import { get } from "../../../utils/theme";
import { mq } from "../../../styles/responsive";

var IconFirst = function IconFirst(p) {
  return !p.opened ? '0px' : '10px';
};

var IconMiddle = function IconMiddle(p) {
  return !p.opened ? '1' : '0';
};

var IconLast = function IconLast(p) {
  return !p.opened ? '0px' : '-6px';
};

var IconRotate = function IconRotate(p) {
  return !p.opened ? '0deg' : '45deg';
};

var Icon = styled.div.withConfig({
  displayName: "Hamburger__Icon"
})(["position:relative;width:23px;height:32px;transform:translateX(", ") translateY(", ") scale(", ");"], function (p) {
  return p.opened ? '-2px' : '-1px';
}, function (p) {
  return p.opened ? '0' : '2px';
}, function (p) {
  return p.opened ? 0.8 : 1;
});
var sidebarBg = get('colors.sidebarBg');
var sidebarPrimary = get('colors.sidebarPrimary');
var sidebarText = get('colors.sidebarText');
var primaryColor = get('colors.primary');
var backgroundColor = get('colors.background');
var textColor = get('colors.text');
var IconLine = styled.span.withConfig({
  displayName: "Hamburger__IconLine"
})(["content:'';display:block;position:absolute;width:100%;height:2px;left:0;right:0;background:", ";transition:transform 0.3s,opacity 0.3s;&:nth-of-type(1){top:-2px;transform:translateY(", ") rotate(", ");}&:nth-of-type(2){top:6px;opacity:", ";}&:nth-of-type(3){top:14px;transform:translateY(", ") rotate(-", ");}"], function (p) {
  return p.opened ? sidebarText(p) : textColor(p);
}, IconFirst, IconRotate, IconMiddle, IconLast, IconRotate);

var translateX = function translateX(p) {
  return !p.opened ? '10px' : '-6px';
};

var translateY = function translateY(p) {
  return !p.opened ? '4px' : '0px';
};

var radii = get('radii');
var ToggleButton = styled.button.withConfig({
  displayName: "Hamburger__ToggleButton"
})(["cursor:pointer;z-index:99;position:absolute;display:flex;align-items:center;justify-content:center;padding:5px 6px;width:33px;height:30px;top:", ";right:", ";transform:translateX(", ") translateY(", ");transition:transform 0.3s;outline:none;border:none;background:", ";border-radius:", ";&:before{position:absolute;content:'';top:-3px;left:0;width:calc(100% + 1px);height:", ";background:", ";}", ";"], function (p) {
  return p.opened ? '3px' : '2px';
}, function (p) {
  return p.opened ? '-39px' : '-27px';
}, translateX, translateY, function (p) {
  return p.opened ? sidebarBg(p) : backgroundColor(p);
}, function (p) {
  return p.opened ? "0 0 ".concat(radii(p), " 0") : "".concat(radii(p));
}, function (p) {
  return p.opened ? '3px' : 0;
}, function (p) {
  return sidebarPrimary(p) || primaryColor(p);
}, mq({
  display: ['block', 'block', 'block', 'none']
}));
export var Hamburger = function Hamburger(_ref) {
  var opened = _ref.opened,
      onClick = _ref.onClick;
  return React.createElement(ToggleButton, {
    opened: opened,
    onClick: onClick
  }, React.createElement(Icon, {
    opened: opened
  }, React.createElement(IconLine, {
    opened: opened
  }), React.createElement(IconLine, {
    opened: opened
  }), React.createElement(IconLine, {
    opened: opened
  })));
};
