// docz-theme-umi/es/components/shared/Sidebar/index.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

import * as React from 'react';
import { Fragment, useState, useEffect } from 'react';
import { useMenus, useWindowSize, usePrevious } from 'docz';
import styled from 'styled-components';
import { Logo } from "../Logo";
import { Search } from "../Search";
import { Menu } from "./Menu";
import { Docz } from "./Docz";
import { Hamburger } from "./Hamburger";
import { get } from "../../../utils/theme";
import { mq, breakpoints } from "../../../styles/responsive";
var sidebarBg = get('colors.sidebarBg');
var sidebarText = get('colors.sidebarText');
var sidebarBorder = get('colors.sidebarBorder');
var Wrapper = styled.div.withConfig({
  displayName: "Sidebar__Wrapper"
})(["position:relative;width:280px;min-width:280px;min-height:100vh;background:", ";transition:transform 0.2s,background 0.3s;z-index:1000;", ";dl{padding:0;margin:0 16px;}dl a{font-weight:400;}@media screen and (max-width:", "px){transform:translateX(", ");position:", ";}", ";"], sidebarBg, mq({
  position: ['absolute', 'absolute', 'absolute', 'relative']
}), breakpoints.desktop - 1, function (p) {
  return p.opened ? '-100%' : '0';
}, function (p) {
  return p.opened ? 'auto' : 'fixed';
}, get('styles.sidebar'));
var Content = styled.div.withConfig({
  displayName: "Sidebar__Content"
})(["position:fixed;top:0;left:0;display:flex;flex-direction:column;width:280px;min-width:280px;height:100%;max-height:100vh;"]);
var Menus = styled.nav.withConfig({
  displayName: "Sidebar__Menus"
})(["flex:1;overflow-y:auto;margin-bottom:10px;"]);
var Empty = styled.div.withConfig({
  displayName: "Sidebar__Empty"
})(["flex:1;opacity:0.7;padding:0 24px;color:", ";"], sidebarText);
var Footer = styled.div.withConfig({
  displayName: "Sidebar__Footer"
})(["padding:10px 0;display:flex;align-items:center;justify-content:center;font-size:14px;color:", ";border-top:1px dashed ", ";"], sidebarText, sidebarBorder);
var FooterLink = styled.a.withConfig({
  displayName: "Sidebar__FooterLink"
})(["padding:0;margin:0;margin-left:5px;"]);
var FooterLogo = styled(Docz).withConfig({
  displayName: "Sidebar__FooterLogo"
})(["fill:", ";"], sidebarText);
var ToggleBackground = styled.div.withConfig({
  displayName: "Sidebar__ToggleBackground"
})(["content:'';display:", ";position:fixed;background-color:rgba(0,0,0,0.4);width:100vw;height:100vh;top:0;bottom:0;left:0;right:0;cursor:pointer;z-index:99;"], function (p) {
  return p.opened ? 'none' : 'block';
});
export var Sidebar = function Sidebar() {
  var _useState = useState(true),
      _useState2 = _slicedToArray(_useState, 2),
      hidden = _useState2[0],
      setHidden = _useState2[1];

  var _useState3 = useState(''),
      _useState4 = _slicedToArray(_useState3, 2),
      query = _useState4[0],
      setQuery = _useState4[1];

  var menus = useMenus({
    query: query
  });
  var windowSize = useWindowSize();
  var isDesktop = windowSize.outerWidth >= breakpoints.desktop;
  var prevIsDesktop = usePrevious(isDesktop);
  useEffect(function () {
    if (!hidden && !prevIsDesktop && isDesktop) {
      setHidden(true);
      document.documentElement.classList.remove('with-overlay');
    }
  });

  var addOverlayClass = function addOverlayClass(isHidden) {
    var method = !isHidden ? 'add' : 'remove';

    if (window && typeof window !== 'undefined' && !isDesktop) {
      document.documentElement.classList[method]('with-overlay');
    }
  };

  var handleSidebarToggle = function handleSidebarToggle() {
    setHidden(function (s) {
      return !s;
    });
    addOverlayClass(!hidden);
  };

  var builtWith = process.env.BIGFISH_VERSION ? ['Bigfish', 'https://bigfish-pre.antfin-inc.com/'] : ['umi-library', 'https://github.com/umijs/umi/tree/master/packages/umi-library'];
  return React.createElement(Fragment, null, React.createElement(Wrapper, {
    opened: hidden
  }, React.createElement(Content, null, React.createElement(Hamburger, {
    opened: !hidden,
    onClick: handleSidebarToggle
  }), React.createElement(Logo, {
    showBg: !hidden
  }), React.createElement(Search, {
    onSearch: setQuery
  }), menus && menus.length === 0 ? React.createElement(Empty, null, "No documents found.") : React.createElement(Menus, null, menus && menus.map(function (menu) {
    return React.createElement(Menu, {
      key: menu.id,
      item: menu,
      sidebarToggle: handleSidebarToggle,
      collapseAll: Boolean(query.length)
    });
  })), React.createElement(Footer, null, "Built with", React.createElement("span", null, "\xA0"), React.createElement("a", {
    href: builtWith[1],
    target: "_blank"
  }, builtWith[0]), React.createElement("span", null, "\xA0"), "and", React.createElement(FooterLink, {
    href: "https://docz.site",
    target: "_blank"
  }, React.createElement(FooterLogo, {
    width: 40
  }))))), React.createElement(ToggleBackground, {
    opened: hidden,
    onClick: handleSidebarToggle
  }));
};
