// docz-theme-umi/es/components/shared/Sidebar/MenuHeadings.js
import * as React from 'react';
import { useDocs, useConfig } from 'docz';
import { Location } from '@reach/router';
import styled from 'styled-components';
import get from 'lodash/get';
import { SmallLink } from "./SmallLink";
var Submenu = styled.div.withConfig({
  displayName: "MenuHeadings__Submenu"
})(["display:flex;flex-direction:column;margin:5px 0 0 24px;"]);

var getHeadings = function getHeadings(route, docs) {
  var doc = docs.find(function (doc) {
    return doc.route === route;
  });
  var headings = get(doc, 'headings');
  return headings ? headings.filter(function (heading) {
    return heading.depth === 2;
  }) : [];
};

export var MenuHeadings = function MenuHeadings(_ref) {
  var route = _ref.route,
      onClick = _ref.onClick;
  var docs = useDocs();

  var _useConfig = useConfig(),
      Link = _useConfig.linkComponent;

  var headings = docs && getHeadings(route, docs);
  return headings && headings.length > 0 ? React.createElement(Location, null, function (_ref2) {
    var location = _ref2.location;
    return React.createElement(Submenu, null, headings.map(function (heading) {
      return React.createElement(SmallLink, {
        as: Link,
        location: location,
        key: heading.slug,
        onClick: onClick,
        to: "".concat(route, "#").concat(heading.slug),
        slug: heading.slug
      }, heading.value);
    }));
  }) : null;
};
